src/dmd/json.d

```diff
@@ -857,17 +880,161 @@ public:
 extern (C++) void json_generate(OutBuffer* buf, Modules* modules)
 {
     scope ToJsonVisitor json = new ToJsonVisitor(buf);
-    json.arrayStart();
-    json.generateCompilerInfo();
-    json.generateBuildInfo();
-    for (size_t i = 0; i < modules.dim; i++)
-    {
-        Module m = (*modules)[i];
-        if (global.params.verbose)
-            message("json gen %s", m.toChars());
-        m.accept(json);
-    }
-    json.generateSemantics();
-    json.arrayEnd();
-    json.removeComma();
+
+    if (global.params.jsonQuery is null)
+    {
+        // Generate the original "non-query format, which is just an array
+        // of modules representing their syntax.
+        json.generateModules(modules);
+        json.removeComma();
+    }
+    else
+    {
+        // Generate the "query format" which is an object where each
+        // output option is it's own field.
+
+        json.objectStart();
+        auto parser = QueryParser(global.params.jsonQuery);
+        while (true)
+        {
+            parser.skipTrivial();
+            if (*parser.next == '\0')
+            {
+                break;
+            }
+            auto fieldName = parser.tryParseName();
+            if (fieldName is null)
+            {
+                json.generateError("invalid query at offset %d: expected name but got '%c' (0x%02x)",
```

Comment:
```
'%c' => `%c`
```

Reply:
I think the single-quotes makes more sense when printing a raw character.

Reply:
Just "name" isn't so helpful, how about generating an up-to-date list?

```d
enum names = () {
    string s;
    static immutable delim = ", ";
    foreach (member; __traits(allMembers, JsonField))
        s ~= member.stringof ~ delim;
    return s[0 .. $ - delim.length];
}();
```

Reply:
The first error is caused by an invalid character, providing all the field names might be too much information in this case.  However, providing the names for the the 2nd error, which is an invalid field name definitely makes sense.  I've included this change for the 2nd error.

Reply:
> I think the single-quotes makes more sense when printing a raw character.

then how about doing like:
src/dmd/parse.d:7425:77:                                error("mismatched string literal postfixes `'%c'` and `'%c'`", postfix, token.postfix);

 the point is it'll get syntax highlighted